[#7985] Remove `msiAclPolicy(...)` and `acAclPolicy` static PEP (main)

MartinFlores751 opened the pull request:
Quick link: #7985.

This PR removes both `msiAclPolicy(...)` and the static PEP `acAclPolicy`.

---

- [x] Remove any remaining `*AclPolicy` bits.
- [x] Check what relies on `STRICT_ACL_KW` and if anything needs to be done about that
- [x] Investigate if `DISABLE_STRICT_ACL_KW` needs any changes
- [x] Investigate if `irods::AGENT_CONN_KW` with `StrictACLOverride` causes an ACL change.
  - [x] See if removing the specific use broke anything.
- [x] Build and test the PR.

korydraughn commented on server/re/src/icatAdminMS.cpp:
```diff
@@ -633,82 +633,6 @@ int msiRenameLocalZoneCollection(msParam_t* _new_zone_name, ruleExecInfo_t* _rei
-        return ret.code();
-    }
-
-    std::string strict = "off";
-    inputArg = ( char * ) msParam->inOutStruct;
-    if ( inputArg != NULL ) {
-        if ( strncmp( inputArg, "STRICT", 6 ) == 0 ) {
-            if( irods::KW_CFG_SERVICE_ROLE_PROVIDER == svc_role ) {
-                chlGenQueryAccessControlSetup( NULL, NULL, NULL, 0, 2 );
-                strict = "on";
-            }
-        }
-    }
-    else {
-        if( irods::KW_CFG_SERVICE_ROLE_PROVIDER == svc_role ) {
-            chlGenQueryAccessControlSetup( NULL, NULL, NULL, 0, 0 );
-        }
-    }
-
-    // set a strict acl prop
-    try {
-        irods::set_server_property<std::string>(irods::STRICT_ACL_KW, strict);
-    } catch ( irods::exception& e ) {
-        irods::log(e);
-        return e.code();
-    }
```
You're going to need to port this into agent_main.cpp (i.e. the agent factory).

Notice the check on L686.

korydraughn replied:
Port it, but without the options to change it from STRICT mode.

MartinFlores751 replied:
Spoiler for a commit I haven't pushed yet (going to right after). I put the following from the function you highlighted at the old call site of the static PEP:
```c++
if (irods::KW_CFG_SERVICE_ROLE_PROVIDER == svc_role) {
        chlGenQueryAccessControlSetup(NULL, NULL, NULL, 0, 2);
}
```

Do we set this up in agent main?

I'm still seeing more calls to `chlGenQueryAccessControlSetup(...)` in the static PEP call-site: https://github.com/irods/irods/blob/a73a90cf0ba46c86f7cb6e9c5eae7e132faeb147/server/api/src/rsGenQuery.cpp#L840-L855

korydraughn replied:
> I put the following from the function you highlighted at the old call site of the static PEP.

Old call site of the static PEP? Got a link to a source file and line number?

> Do we set this up in agent main?

`rsGenQuery` is invoked quite early in the process and is the first function which triggers the PEP. I suspect that's why you're seeing multiple invocations of `chlGenQueryAccessControlSetup`.

MartinFlores751 replied:
> > I put the following from the function you highlighted at the old call site of the static PEP.
> 
> Old call site of the static PEP? Got a link to a source file and line number?

Here it is: https://github.com/irods/irods/blob/a73a90cf0ba46c86f7cb6e9c5eae7e132faeb147/server/api/src/rsGenQuery.cpp#L808-L834

> > Do we set this up in agent main?
> 
> `rsGenQuery` is invoked quite early in the process and is the first function which triggers the PEP. I suspect that's why you're seeing multiple invocations of `chlGenQueryAccessControlSetup`.

Yeah, kinda figured. Not sure what all the magic numbers at the end mean, so that's why I put the call at the old call site.

korydraughn replied:
I think all of that can be removed in favor of setting the necessary values in agent_main.cpp. There's no reason to have GenQuery1 be responsible for that anymore.

korydraughn replied:
Moving it to the agent factory's init logic means it will be invoked one time too.

korydraughn replied:
Basically, somewhere before this point.

https://github.com/irods/irods/blob/a73a90cf0ba46c86f7cb6e9c5eae7e132faeb147/server/main_server/src/agent_main.cpp#L300

MartinFlores751 replied:
Can the `irods::KW_CFG_SERVICE_ROLE_PROVIDER == svc_role` check be done at that point?

korydraughn replied:
It should be possible to do the check there.

MartinFlores751 replied:
Will try and then take out of draft for building and testing.

MartinFlores751 replied:
Done.

MartinFlores751 commented:
Building and testing.

alanking commented on server/api/src/rsGenQuery.cpp:
```diff
@@ -760,8 +760,6 @@ _rsGenQuery( rsComm_t *rsComm, genQueryInp_t *genQueryInp,
     int resc_hier_attr_pos = -1;
 
     static int ruleExecuted = 0;
```
Can `ruleExecuted` also be removed?

MartinFlores751 replied:
Add check with `ruleExecuted` back in.

alanking commented on server/main_server/src/agent_main.cpp:
```diff
@@ -297,6 +297,13 @@ auto main(int _argc, char* _argv[]) -> int
         irods::re_plugin_globals->global_re_mgr.call_setup_operations();
 
+        std::string svc_role;
+        irods::error ret = get_catalog_service_role(svc_role);
```
Do we need to handle `!ret.ok()` here?

MartinFlores751 replied:
Done.

alanking commented on server/re/src/icatAdminMS.cpp:
```diff
@@ -633,82 +633,6 @@ int msiRenameLocalZoneCollection(msParam_t* _new_zone_name, ruleExecInfo_t* _rei
-
-    // set a strict acl prop
-    try {
-        irods::set_server_property<std::string>(irods::STRICT_ACL_KW, strict);
```
Not sure if we're allowed to remove it, but I see `DISABLE_STRICT_ACL_KW`. Should stop supporting this in the server and/or remove it, or at least check to see how it's being used?

trel replied:
yes, please let's not support something with that name

MartinFlores751 replied:
Removed.

alanking reviewed (commented):
Looking good so far. Saw a couple other things that we might be able to chop out

MartinFlores751 commented:
Unit tests failed.

Have a `CHECK( rc_atomic_apply_acl_operations(static_cast&lt;RcComm*>(conn), json_input.c_str(), &amp;json_error_string) == ec )` which expands to `-830000 == -358000`.

There are some other errors too.

MartinFlores751 commented on server/api/src/rsGenQuery.cpp:
```diff
@@ -805,31 +810,40 @@ _rsGenQuery( rsComm_t *rsComm, genQueryInp_t *genQueryInp,
+            constexpr auto control_flag{2};
+
+            // Call directly to avoid additional policy
+            chl_gen_query_access_control_setup_impl(nullptr, nullptr, nullptr, priv, control_flag);
```
Building with a forward declared `chl_gen_query_access_control_setup_impl` doesn't work as is right now. Going to use `chlGenQueryAccessControlSetup` for now to tackle `DISABLE_STRICT_ACL_KW`, will revisit this afterward.

korydraughn replied:
That's very surprising. Anyway, that can be addressed later.

Carry on.

MartinFlores751 replied:
I'm not 100% sure, but I think the reason that this doesn't work is that `general_query.cpp`, where `chl_gen_query_access_control_setup_impl` is defined, is not being linked (or set as a dependency?) against the `irods_server_api` target in CMake.

Will resolve if we're fine with using `chlGenQueryAccessControlSetup` still.

MartinFlores751 replied:
In standup: We decided that we're going to create an issue for using `chl_gen_query_access_control_setup_impl` and not use it in this PR to get it moving.

MartinFlores751 replied:
Created #8304.

MartinFlores751 commented:
Found this bit of code: https://github.com/irods/irods/blob/f11868c488e1eec26e9cce0af227dd0f860d9fbd/server/core/src/dataObjOpr.cpp#L820-L833

Is `irods::AGENT_CONN_KW` used anywhere outside of the server to make some decisions? Do they check for `StrictACLOverride` as the value?

---

I think this is used in the irods server in combination with the `DISABLE_STRICT_ACL_KEYWORD`, going to check for that...

MartinFlores751 commented:
Rebased onto main, I think the tests were failing due to some other tests. Updated the icommands companion PR as well.

Rebuilding and rerunning tests.

alanking commented on server/api/src/rsGenQuery.cpp:
```diff
@@ -805,54 +794,50 @@ _rsGenQuery( rsComm_t *rsComm, genQueryInp_t *genQueryInp,
-    char* dis_kw = getValByKey( &genQueryInp->condInput, DISABLE_STRICT_ACL_KW );
-    if ( agent_conn_flag && dis_kw ) {
-        acl_val = 0;
+            // Call directly to avoid additional policy
```
This call is still going to trigger at least one database PEP (`pep_database_gen_query_access_control_setup_*`), so I'm not sure that I understand this comment. Please explain.

MartinFlores751 replied:
That comment was associated with this change: https://github.com/irods/irods/pull/8264#discussion_r1994092194

It was reverted since the code wouldn't compile. Will remove if we continue to use `chlGenQueryAccessControlSetup(...)`.

MartinFlores751 replied:
Removed.

alanking reviewed (commented):
Seems good on inspection. Hoping the tests pass

MartinFlores751 commented:
Single test failure: `irods.test.test_native_rule_engine_plugin.Test_Native_Rule_Engine_Plugin.test_network_pep`. Multiple attempts to run fail.

All other unit and core tests pass.

Short log:
```
Traceback (most recent call last):
  File "/var/lib/irods/scripts/irods/test/test_native_rule_engine_plugin.py", line 543, in test_network_pep
    self.helper_test_pep(pep_map[self.plugin_name], "iput -f --metadata ATTR;VALUE;UNIT "+self.testfile)
  File "/var/lib/irods/scripts/irods/test/test_native_rule_engine_plugin.py", line 123, in helper_test_pep
    check_string_count_in_log_section (s,number_of_strings_to_look_for)
  File "/var/lib/irods/scripts/irods/test/test_native_rule_engine_plugin.py", line 112, in check_string_count_in_log_section
    lib.delayAssert(
  File "/var/lib/irods/scripts/irods/lib.py", line 552, in delayAssert
    raise AssertionError
AssertionError
```

alanking commented:
> Single test failure: `irods.test.test_native_rule_engine_plugin.Test_Native_Rule_Engine_Plugin.test_network_pep`. Multiple attempts to run fail.

I have been seeing this test fail occasionally on main as well. What OS / database are you using? I'll write up an issue for this and we can move this along.

alanking commented on server/api/src/rsGenQuery.cpp:
```diff
@@ -805,54 +794,50 @@ _rsGenQuery( rsComm_t *rsComm, genQueryInp_t *genQueryInp,
-                           rsComm->clientAddr,
-                           rsComm->clientUser.authInfo.authFlag,
-                           acl_val );
+    int old_acl_val = chlGenQueryAccessControlSetup(rsComm->clientUser.userName,
```
Codacy says... `old_acl_val` is never used (which is true).

MartinFlores751 replied:
Done.

MartinFlores751 commented:
Reverting the most recent commit has all tests passing, with two retries on one test.

`irods.test.test_ichmod.Test_ichmod.test_ichmod_access_levels` failed twice with the following stack trace:

```
Traceback (most recent call last):
  File "/var/lib/irods/scripts/irods/test/test_ichmod.py", line 208, in test_ichmod_access_levels
    self.user.assert_icommand(params.get('cmd'), *check[2])
  File "/var/lib/irods/scripts/irods/test/session.py", line 162, in assert_icommand
    return assert_command(*args, **kwargs)
  File "/var/lib/irods/scripts/irods/test/command.py", line 76, in assert_command
    return _assert_helper(*args, should_fail=False, **kwargs)
  File "/var/lib/irods/scripts/irods/test/command.py", line 104, in _assert_helper
    assert result
AssertionError
```

The specific line that failed:
```
...
 --- IrodsSession: icommand executed by [alice#tempZone] [irmdir /tempZone/home/public/test_ichmod_collection_access] --- 
Assert Command: irmdir /tempZone/home/public/test_ichmod_collection_access
Expecting EMPTY: ['']
  stdout:
    | rmdirColl: rcRmColl failed with error -818000
  stderr:
    | 
Unexpected output on stdout

FAILED TESTING ASSERTION
...
```

alanking commented:
Thanks. Created https://github.com/irods/irods/issues/8299 to address the occasional failure for `test_native_rule_engine_plugin.Test_Native_Rule_Engine_Plugin.test_network_pep`.

As for `test_ichmod.Test_ichmod.test_ichmod_access_levels`, that one is quite unfortunate to not have employed a `subTest` so we can see what failed. I don't like that it failed and then started working after changing nothing. :) Please run a few more times in isolation and see if things improve.

MartinFlores751 commented:
> As for `test_ichmod.Test_ichmod.test_ichmod_access_levels`, that one is quite unfortunate to not have employed a `subTest` so we can see what failed. I don't like that it failed and then started working after changing nothing. :) Please run a few more times in isolation and see if things improve.

This one behaved like this after reverting the latest commit. Still, it'll be tested again after addressing the remaining comments.

MartinFlores751 commented:
> Thanks. Created #8299 to address the occasional failure for `test_native_rule_engine_plugin.Test_Native_Rule_Engine_Plugin.test_network_pep`.
> 
> As for `test_ichmod.Test_ichmod.test_ichmod_access_levels`, that one is quite unfortunate to not have employed a `subTest` so we can see what failed. I don't like that it failed and then started working after changing nothing. :) Please run a few more times in isolation and see if things improve.

Rebuilding and retesting.

MartinFlores751 commented:
Core and unit tests passed.

korydraughn commented:
Good. The question to answer now is - is a run of the unit/core tests enough to prove correctness?

Please review the documentation for STRICT to determine whether we can put other a test which bangs on the change directly.

MartinFlores751 commented:
The direct documentation doesn't help much:

https://github.com/irods/irods_docs/blob/c9f5c10479e090c79c0517f36fc5c33d1da8d419/docs/system_overview/users_and_permissions.md?plain=1#L15-L37

But the changelogs show things that were/should be affected by the ACLs, here's one example:

https://github.com/irods/irods_docs/blob/c9f5c10479e090c79c0517f36fc5c33d1da8d419/docs/legacy_release_notes.md?plain=1#L93

---

I think it might be worth doing another check of what calls the init function, and figuring out what the magic numbers really mean, to make sure the fix  is correct. Put differently, making sure that whatever the magic input is (unless there's a special exception), the 'state' stays 'strict'.

I'll probably do this in a bit, unless others think it's unnecessary.

korydraughn commented:
After doing some review with @MartinFlores751, we've agreed to revert the changes which remove the `DISABLE_STRICT_ACL_KW` keyword/usage. We're going to leave it in until we have a better understanding of how it's used.

For now, our goal is to focus on disallowing modification of the ACL policy.

MartinFlores751 commented:
Rebase onto tip of main.

MartinFlores751 commented:
Building and running tests.

MartinFlores751 commented:
Unit tests passed but core tests failed. Only `test_all_rules.Test_AllRules` failed.

Looking into the logs, we see:
```
 --- IrodsSession: icommand executed by [rods#tempZone] [irule -vF /var/lib/irods/clients/icommands/test/rules/rulemsiAclPolicy.r -r irods_rule_engine_plugin-irods_rule_language-instance] --- 
Assert Command: irule -vF /var/lib/irods/clients/icommands/test/rules/rulemsiAclPolicy.r -r irods_rule_engine_plugin-irods_rule_language-instance
Expecting STDOUT_SINGLELINE: ['completed successfully']
  stdout:
    | 
  stderr:
    |  ERROR: Incomplete rule input for /var/lib/irods/clients/icommands/test/rules/rulemsiAclPolicy.r
Unexpected output on stderr

FAILED TESTING ASSERTION
```

Whichs shows that I forgot to use the companion icommands PR... Which I checked and is true.

Will rerun that specific test with the companion PR just to see it pass.

MartinFlores751 commented:
`test_all_rules.Test_AllRules` passed with the companion PR.

alanking reviewed (commented):
Squash em and will take one last pass over it

MartinFlores751 commented:
Squashed.

alanking reviewed (approved):
\# it

MartinFlores751 commented:
Pounded.
